src/noit_metric_director.c

```diff
@@ -0,0 +1,231 @@
+  }
+  return my_lane.id;
+}
+
+void
+noit_adjust_metric_interest(uuid_t id, const char *metric, short cnt) {
+  int thread_id, icnt;
+  void *vhash, *vinterests;
+  mtev_hash_table *level2;
+  caql_cnt_t *interests;
+
+  thread_id = get_my_lane();
+
+  /* Get us our UUID->HASH(METRICS) mapping */
+  if(!mtev_hash_retrieve(&id_level, (const char *)id, UUID_SIZE, &vhash)) {
+    int found;
+    uuid_t *copy = malloc(UUID_SIZE);
+    level2 = calloc(1,sizeof(*level2));
+    uuid_copy(*copy, id);
+    if(!mtev_hash_store(&id_level, (const char *)copy, UUID_SIZE, level2)) {
+      free(copy);
+      free(level2);
+    }
+    found = mtev_hash_retrieve(&id_level, (const char *)id, UUID_SIZE, &vhash);
+    assert(found);
+  }
+  level2 = vhash;
+
+  if(!mtev_hash_retrieve(level2, metric, strlen(metric), &vinterests)) {
+    int found;
+    char *metric_copy;
+    metric_copy = strdup(metric);
+    vinterests = calloc(nthreads, sizeof(*interests));
+    if(!mtev_hash_store(level2, metric_copy, strlen(metric_copy), vinterests)) {
+      free(metric_copy);
+      free(vinterests);
+    }
+    found = mtev_hash_retrieve(level2, metric, strlen(metric), &vinterests);
+    assert(found);
+  }
+  interests = vinterests;
+  /* This is fine because thread_id is only ours */
+  icnt = interests[thread_id];
+  icnt += cnt;
+  if(icnt < 0) icnt = 0;
+  assert(icnt <= 0xffff);
+  interests[thread_id] = icnt;
+}
+
+static void
+distribute_metric(struct metric *metric, 
+                  const char *line, int line_len) {
+  int i;
+  void *vhash, *vinterests;
+  char uuid_str[UUID_STR_LEN+1];
+  uuid_unparse_lower(metric->id, uuid_str);
+  if(mtev_hash_retrieve(&id_level, (const char *)&metric->id, UUID_SIZE, &vhash)) {
+    if(mtev_hash_retrieve((mtev_hash_table *)vhash,
+                          metric->metric_name, metric->metric_name_len,
+                          &vinterests)) {
+      caql_cnt_t *interests = vinterests;
+      for(i=0;i<nthreads;i++) {
+        if(interests[i] > 0) {
+          char *copy = mtev__strndup(line, line_len);
+          ck_fifo_spsc_t *fifo = (ck_fifo_spsc_t *)thread_queues[i];
+          ck_fifo_spsc_entry_t *fifo_entry;
+          fifo_entry = ck_fifo_spsc_recycle(fifo);
+          if(!fifo_entry) fifo_entry = malloc(sizeof(ck_fifo_spsc_entry_t));
+          ck_fifo_spsc_enqueue_lock(fifo);
+          ck_fifo_spsc_enqueue(fifo, fifo_entry, copy);
+          ck_fifo_spsc_enqueue_unlock(fifo);
+        }
+      }
+    }
+  }
+}
+
+char *noit_metric_director_lane_next() {
+  char *line = NULL;
+  if(my_lane.fifo == NULL) return NULL;
+  ck_fifo_spsc_dequeue_lock(my_lane.fifo);
+  if(ck_fifo_spsc_dequeue(my_lane.fifo, &line) == false) {
+    line = NULL;
+  }
+  ck_fifo_spsc_dequeue_unlock(my_lane.fifo);
+  return line;
+}
+static void
+handle_metric_buffer(const char *payload, int payload_len, int has_noit) {
+  switch(payload[0]) {
+    case 'H':
+    case 'M':
+      {
+        struct metric metric;
```

Comment:
Is the confusion specific to the variable being named the same thing as the struct?